### PR TITLE
Added support for Gemma-2

### PR DESCRIPTION
**Pull request by neelnanda-io**
Added support for Gemma-2 models.

- Support for attention and output logit soft-capping (ie attn_scores -> 50 * tanh(attn_scores / 50), basically squashing it into the interval [-50, 50])
  - Gotcha: Output logits with softcap are *not* invariant to adding a constant.
- Support for before and after LN
- Minor refactors, eg now MLP activation functions are defined as a dict in utils.py rather than a stack of if statements in both mlp.py and gated_mlp.py

Key differences between Gemma-1 and Gemma-2:
* Gemma-2 has an RMS Norm at the start *and* end of each layer. You cannot fold the RMS Norm weights of the ones at the end.
* Attn_scores and final logits are softcapped (with constants 50 and 30 respectively)
* Every other attn layer (1, 3, 5, etc) uses local, sliding window attention, with window 4096 (the max ctx is 8192, so I don't know why they bother tbh)
* They use grouped query attention, with group size 2 (ie every adjacent pair of heads use the same value and key)

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# Checklist:

- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- **Note**: I have not manually updated the model_properties_table - is this automatic or do I need to?
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- Given that the smallest model is 9B, tests don't really seem practical. I have manually verified that the logits agree with HuggingFace for the 9B on the string "<BOS>The capital of France is", at max deviation 8.2016e-05 for float32, though notably worse for low precision: 0.5 for bfloat16, 0.0664 for float16. To me this seems annoying but acceptable. I have not tested that things work well on really long strings or generations.
- [ ] New and existing unit tests pass locally with my changes
- [X] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Comment by neelnanda-io** [comment]
Looks like abstract_attention.py assumed that n_heads * d_head == d_model, which is normally true but not true for Gemma-2 27B, I fixed that in the latest commit (this should do nothing for any model where that relation is true)

**Comment by neelnanda-io** [comment]
Hmm, there's at least a 0.2 difference in logits for the 27B in float32, which is concerning... And quite surprising tbh, I don't see any architectural difference between the 9B and 27B, which suggests this would only be cascading errors? I did get the HF logits on CPU and TL logits across 2 GPUs though, which might cause some additional divergence?

**Comment by ArthurConmy** [comment]
> Hmm, there's at least a 0.2 difference in logits for the 27B in float32, which is concerning... And quite surprising tbh, I don't see any architectural difference between the 9B and 27B, which suggests this would only be cascading errors? I did get the HF logits on CPU and TL logits across 2 GPUs though, which might cause some additional divergence?

I think you should move on and put a loud warning when loading 27B model. We have lots of evidence that TransformerLens' slight differences from HuggingFace have cascading errors, e.g. [here](https://github.com/TransformerLensOrg/TransformerLens/issues/615) and [here](https://github.com/TransformerLensOrg/TransformerLens/issues/570). @bryce13950 has mentioned trying to improve the numerical errors, so I doubt you made the error in implementation. The only question to me is whether the 27B model should even be merged.

**Comment by neelnanda-io** [comment]
Fair. I think 27B should be merged, but printing a warning sounds good. I was mostly surprised at such a big jump between 9B and 27B of the error, when even cascading errors shouldn't explain that IMO - it's like 42 vs 46 layers, just a fair bit wider and with about twice as many neurons per MLP layer

**Comment by bryce13950** [comment]
I am going to try and fold in some accuracy improvements, specifically in MLPs, into this when I put it up. I wouldn’t worry about adding warnings or anything for the being. I have a list of models to try once that is done. This is second on the list now.

**Comment by bryce13950** [comment]
We crossed paths a little bit with this. I have done quite a bit with MLPs recently, and we did a few things very similar. I have been thinking about how to put the two changes together, and I think I am going to wrap up my branch first, given that it affects accuracy with existing models. I am basically redoing the entire set of components, and I would bet that what I am doing is going to increase accuracy here.

File: transformer_lens/components/transformer_block.py [hunk 1]
```diff
@@ -157,21 +168,33 @@ def forward(
...
             resid_mid = self.hook_resid_mid(resid_pre + attn_out)  # [batch, pos, d_model]
             mlp_in = (
                 resid_mid if not self.cfg.use_hook_mlp_in else self.hook_mlp_in(resid_mid.clone())
             )
             normalized_resid_mid = self.ln2(mlp_in)
-            mlp_out = self.hook_mlp_out(self.mlp(normalized_resid_mid))  # [batch, pos, d_model]
+            mlp_out = self.mlp(normalized_resid_mid)  # [batch, pos, d_model]
+            if self.cfg.use_normalization_before_and_after:
+                mlp_out = self.ln2_post(mlp_out)
+            mlp_out = self.hook_mlp_out(mlp_out)
```
Comment (bryce13950) [inline transformer_lens/components/transformer_block.py:186]: Can we break this out into a function? Not a fan of repeated logic, as it is often where bugs can happen either through things like initial typos, or someone changing something down the line without realizing the logic exists twice. Maybe a function named `apply_mlp` that we can call here, and then move this block into that.

**Comment by bryce13950** [comment]
Just one comment from me on the code so far. I am going to wrap up my work, and come back to this afterwards to test it, and load up the code locally to play around a bit.

File: transformer_lens/components/transformer_block.py (hunk 1 above)
Comment (neelnanda-io) [reply transformer_lens/components/transformer_block.py:186]: The repeated logic being that this is done after attn _and_ after mlp? I don't see a good way to combine it in abstract logic, since they use different LNs (I could make a 'apply_normalisation_after' function that takes as input the output and whether it's attn or mlp, and chooses the right LN accordingly, but that's kinda ugly and hard to read IMO.

We could pull it out of the forward pass into a separate fn, but I don't think this would do much to minimise risk if there's a separate apply_mlp and apply_attn function.

Comment (bryce13950) [reply transformer_lens/components/transformer_block.py:186]: I am referring to this line, and line 194. They are currently identical outside of the input variable.

Comment (neelnanda-io) [reply transformer_lens/components/transformer_block.py:186]: Oh, gotcha. Yeah that seems reasonable to me

**Comment by JThh** [comment]
Hey @neelnanda-io thanks for making this pr to support Gemma2 series model. I am aware that this pr has not been ready, but I use what it has and am training a SAE on the `resid_post` position. The run is being recorded [here](https://wandb.ai/jiatongg/sae_semantic_entropy/runs/e88i5gcc?nw=nwuserjiatongg). It might also be good auxiliary reference for further code adjustments to this pr.

**Comment by neelnanda-io** [comment]
Cool! Let me know if you run into any issues

**Comment by bryce13950** [comment]
MLP outputs are now perfect on these models

**Comment by bryce13950** [comment]
image
